PsychPython/audio.py

```diff
@@ -226,11 +231,14 @@ def get_audio_data(self, secs_allocate=None,
                               secs_allocate, min_secs, max_secs,
                               single_type)
 
-    def fill_buffer(self, data):
+    def fill_buffer(self, data, start_index='append'):
         """Create a new buffer and fill with audio data to be played when
         stream starts"""
-        buffer = Buffer(stream=self, data=data)
-        return buffer
+        if self.buffer:
+            self.buffer.fill_buffer(data=data, start_index=start_index)
+        else:
+            self.buffer = Buffer(stream=self, data=data)
+        return self.buffer
```

Comment:
I think there's a misunderstanding on how these buffers work, so above code won't quite work this way:

'FillBuffer' always fills or refills (or streaming refills, e.g., via append) the "builtin" default audio buffer of a stream - often also referenced as special buffer handle 0. It can't create a buffer, so the below self.handle = PsychPortAudio('FillBuffer',...) would not work, as no buffer handle is returned, only the default builtin buffer is filled.

'CreateBuffer' creates a named buffer with a specific returned buffer handle.
'RefillBuffer' is like 'FillBuffer' but works on both named buffers created via 'CreateBuffer' and the default buffer handle 0 -- ie. turns into 'FillBuffer' for special handle 0.

Historically, earlier versions of PsychPortAudio only had the default builtin buffer (0) per stream.
Later on i added those individual named buffers 'Createbuffer'/'DeleteBuffer'/'Refillbuffer', so one can pregenerate many sound buffers, e.g., during experiment startup, and then more quickly switch between them for playback. E.g., by using 'FillBuffer' and passing a buffer handle self.handle instead of a data matrix -- basically attaching that buffer to the playback stream for playback.
The bigger motivator for those named buffers is the sound schedule functionality 'UseSchedule' / 'AddToSchedule' etc. to define "audio playlists" of different buffers, where you can set individual start/stop times, transition times, pauses, repeat counts etc. for sequences of individual buffers, or append new buffers on the fly to an already running audio schedule.

BasicSoundScheduleDemo and BasicAMAndMixScheduleDemo illustrate use of sound schedules, named buffers and also audio slaves and mixing. DelayedSoundFeedbackDemo is a good one to demonstrate streaming refill / append functionality and such with just the single builtin default buffer for audio delay feedback loops and such. GNU/Octave is a good way to run these demos (with the exception of macOS where it still runs, but Apple introduced so many new incompatibilities and breakage into their recent macOS that Homebrew Octave's GUI doesn't work well at all, on at least macOS 10.13 and later).

Reply:
> I think there's a misunderstanding on how these buffers work,

You're not wrong. It's super-confusing to get one's head around PTB's system of slaves and buffers! ;-) From your demo it looked like one could choose to create-and-fill a buffer in one go, or create first and fill later. So that's the functionality I was trying to achieve with that class.

At the PsychoPy end I use 
- 1 stream
- many slaves, 1 slave per sound object
- 1 buffer per slave

so the issues around multiple buffers per slave, and the errors in what that code appears to do, haven't cropped up.

Reply:
Yep, create & fill in one go is what 'CreateBuffer' does - and returns a buffer handle to it.
You can refill that buffer, or append to it etc. via 'RefillBuffer' as it accepts both a buffer handle and a new data matrix. Or 'DeleteBuffer' the buffer and create a new one if that's preferred.
If you want to just play such a buffer then you can use 'FillBuffer' and pass in the buffer handle instead of a data matrix. If you need to build complex schedules with things like play buffer A once, then buffer B for 2 repetitions, then a break of x seconds, then buffer C, then wait until a specific time before playing buffer D etc. then you can use the schedule functionality to add/queue up many of these buffers.

The slaves and masters are separate from that. A master represents an actual audio device (= an actual audio stream opened in the underlying portaudio library = usually one physical sound card). First there were only masters. Then i added slaves, which are purely virtual sound devices -- they are simulated on top of actual portaudio master devices. Essentially all slaves attached to a master are processed by a background thread and treated individually as iff they were real soundcards - you can attach buffers, schedules, set volumes, start/stop them with individual timings, get timestamps etc. But the background thread takes the virtual "output" of all active slave devices at a given time t and mixes them together into actual output sent to the real underlying portaudio stream of the master device, ie., to the actual sound card. That's btw. why the slaves must have the same samplerate as the master, because they are only simulated on top of one master device via mixing.

It gives a lot of flexibility, but can be quite a bit confusing. The best way to use it is to install Octave and play with our sound demos to get a feel of how one can use it. The help texts are not very good at giving a conceptual overview. Also all the numeric flags don't help -- i never found the time to add symbolic flags which carry more meaning.

I think your mapping makes sense as a starter. 1 stream per physical sound card. Many slaves for different sound objects iff those objects should operate independently with possibly different audio content, sound onset/offset times, volumes, potentially amplitude modulation "envelopes", potentially outputting to different sound channels of the physical sound card etc. 1 buffer per slave/audio object makes sense for a starter. You could either use the default builtin buffer, just using 'FillBuffer' for that, or create a buffer via 'CreateBuffer' to create and assign sound in one go and then 'RefillBuffer' to refill with different content. I think that's what you are aiming at atm. Use of things like audio schedules (one individual schedule for each slave/audio object) would need use of multiple buffers per audio object.

Reply:
Yes, my understanding of most of that was correct. Just the distinction between FIllBuffer and RefillBuffer had me lost!

Having scheduling of multiple sounds within a slave is something I'll leave for another day I think. I tend to implement things when the need arises and the need didn't arise for that one yet!

The next thing to handle will be streaming, usually needed for users that load long movie files and a 1hr 44kHz stream doesn't fit in RAM! I'm sure I would have come across my RefillBuffer issue then but for now the PsychoPy simplified implementation is working fine.